### PR TITLE
[Bug fix] binary_ng: bound the sharded FPU batch by the fp32 DST capacity

### LINKED ISSUE #56958: binary_ng: mixed FP32/BF16 sharded add/subtract fail across tile batches on WH and BH
### Problem

Mixed FP32/BF16 height-sharded `ttnn.add` and `ttnn.subtract` produce incorrect results with fused ABS input activations. Found while validating #56850; this is separate from the SrcA format-state fix in #56738/#56850.

On real WH and BH, 12 cases fail on the unfixed baseline as well as the candidate: two operations × two dtype orders × activation on LHS/RHS/both. This establishes pre-existing failures, not necessarily identical erroneous values (the unfixed baseline also has the separate format-state bug).

### Reproduction

Use a real device and the following pytest test with the repository's `device` fixture:

```python
import pytest
import torch
import ttnn

@pytest.mark.parametrize("op_name", ["add", "subtract"])
@pytest.mark.parametrize("dtypes", [(ttnn.float32, ttnn.bfloat16), (ttnn.bfloat16, ttnn.float32)])
@pytest.mark.parametrize("side", ["lhs", "rhs", "both"])
def test_sharded_mixed_activation(device, op_name, dtypes, side):
    shape = (1, 1, 96, 96)
    grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(0, 0))})
    memory = ttnn.create_sharded_memory_config(
        shape, ttnn.CoreGrid(y=1, x=1), ttnn.ShardStrategy.HEIGHT, ttnn.ShardOrientation.ROW_MAJOR
    )
    tensors = [
        ttnn.from_torch(
            ((torch.arange(9216) * 17 + offset) % 101 - 50).float().reshape(shape) / 8,
            dtype=dtype, layout=ttnn.TILE_LAYOUT, device=device, memory_config=memory,
        )
        for dtype, offset in zip(dtypes, (3, 13))
    ]
    a, b = tensors
    ah, bh = [ttnn.to_torch(t).float() for t in tensors]
    activation = [ttnn.UnaryWithParam(ttnn.UnaryOpType.ABS)]
    if side in ("lhs", "both"):
        ah = ah.abs()
    if side in ("rhs", "both"):
        bh = bh.abs()
    out = getattr(ttnn, op_name)(
        a, b, dtype=ttnn.bfloat16, memory_config=memory, sub_core_grids=grid,
        fast_and_approximate_mode=True,
        input_tensor_a_activations=activation if side in ("lhs", "both") else [],
        input_tensor_b_activations=activation if side in ("rhs", "both") else [],
    )
    expected = ah + bh if op_name == "add" else ah - bh
    torch.testing.assert_close(ttnn.to_torch(out).float(), expected, rtol=0, atol=0)
```

### Investigation so far

- WH investigation identified an FP32 FPU batch-capacity mismatch: an eight-tile batch can exceed the four-tile active destination region. A local diagnostic reducing the batch cap was tested separately; it is not part of #56850.
- BH p100a: baseline `3c5598cdb2896c1e4212b1a106cfbfe57cfbcb31` and candidate `69d22588b0402ae287e5eea7bf44c1e1a2cafbce` each fail all 12 cases, using matching host builds and isolated kernel caches.
- One BH candidate case has exactly 5120/9216 wrong elements: five of nine tiles, with the first four correct. This is consistent with the WH capacity explanation, but **the mechanism and diagnostic fix have not been verified on BH**.
- Expected: all tiles agree with the quantized-input PyTorch reference. Validate the eventual fix on both architectures, around the four/eight-tile boundaries and across activation placement.

### PR DESCRIPTION
**Pull request by Aswinmcw**
### PR Category
Bug fix

### Summary
Fixes #56958

Sharded `binary_ng` FPU ops process a batch of tiles per DST acquire. `binary_ng_program_factory.cpp` set that batch to a flat 8, but `fp32_dest_acc_en` is true whenever any operand is FLOAT32, and under fp32 dest accumulation DST holds only **4** tiles in half-sync mode. The 8-tile batch wrapped past the active half. Same-dtype inputs happened to survive the overspill (each batch is fully serialised by acquire/commit/wait/release); mixed fp32/bf16 inputs did not.

Measured on Blackhole P150b, `main`, height-sharded 96×96 (9 tiles), `dtype=bf16`, FPU path:

| Inputs | Activation | Result on `main` |
|---|---|---|
| bf16/bf16, fp32/fp32 | any | correct |
| fp32/bf16, bf16/fp32 | none | **tile 4 wrong** (1024/1024 elements for `add`) |
| fp32/bf16, bf16/fp32 | LHS | **tiles 0–3 wrong**, 4–8 right |
| fp32/bf16, bf16/fp32 | RHS / both | all tiles wrong — that's the SrcA format-state bug (#56738), fixed separately in #56724/#56850 |

So this isn't an activation bug: plain `ttnn.add(bf16_sharded, f32_sharded)` is wrong on `main`. Tile 4 is exactly the fp32 DST half boundary.

**Fix:** bound the FPU batch by `fp32_dest_acc_en ? 4 : 8`, the rule layernorm, softmax, sdpa and the other fp32 ops already use. `fp32_dest_acc_en` is hoisted above the batch-size selection so a single definition feeds both the batch size and the compute config; the `c_3`/`c_4` intermediate CB sizes already derive from `num_tiles_per_cycle`. SFPU path (batch 2) is unaffected.

### Notes for reviewers
- New `test_binary_ng_sharded_fp32_batch.py`: the issue's shape with no-activation and LHS-activation cases, plus a sweep over 4/5/8/9/12/17-tile shards × {f32/bf16, bf16/f32, f32/f32} so both the full-batch and remainder paths see every batch size. **Fails 20/44 on `main`, passes 44/44 with the fix.** RHS-activation cases are left to #56724/#56850's tests since they need that fix too; with both fixes applied locally all 16 mixed configurations from the issue pass.
- Regression on P150b: `test_binary_ng_bcast_fp32_dest_acc` (55), `test_binary_ng_typecast` (1543), `test_binary_ng_program_cache` (34 + 1 xfail), `test_binaryng_fp32` (64), `test_binary_fp32` (45), `test_binary_dtype_validation` (66), `test_typecast_sharded` (151), `test_add` (102), `test_binary_bcast` (3848) — all pass.
- Cost: fp32 sharded FPU ops go from 8 to 4 tiles per acquire/commit. That's the batch size every other fp32 op uses; the 8-tile version was relying on out-of-spec DST wraparound.
- Verified on Blackhole (P150b). Not run on Wormhole; the issue reports the same failures on WH and the fix is arch-independent (the 4-tile fp32 DST bound is the same on both), but it should be confirmed there. The issue's p100a observation ("first four tiles correct, five of nine wrong") matches the same 4-tile boundary.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:aswinmcw/binary-ng-fp32-dst-batch)

### CODE REVIEW

**Review by KalaivaniMCW — APPROVED** [review]
LGTM

**Review by VirdhatchaniKN — APPROVED** [review]
LGTM
